### ISSUE TITLE
[no-empty-function] private constructors should not error

### ISSUE THREAD

**Issue by jschaf** [typescript-eslint/typescript-eslint#1178]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-empty-function": "on",
  }
}
```

```TS
class Foo {
  private constructor() {}
}
```

**Expected Result**

No warning because a private no-args constructor is useful to force instantiation of a class from static methods.

**Actual Result**

The rule warns that the private constructor is useless.
```
error  Unexpected empty constructor  @typescript-eslint/no-empty-function
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.6.1` |
| `@typescript-eslint/parser`        | `2.6.1` |
| `TypeScript`                       | `3.6.3` |
| `ESLint`                           | `6.6.0` |
| `node`                             | `12.10` |
| `npm`                              | `6.11.3` |

**Comment by a-tarasyuk** [comment]
@bradzacher What is an acceptable option name for this case?

**Comment by bradzacher** [comment]
`allowEmptyPrivateConstructor`?
Generally with eslint options, just aim for clarity. As people configure them once, length/verbosity isn't something to worry about.

**Comment by a-tarasyuk** [comment]
I've checked `no-empty-function` [options](https://eslint.org/docs/rules/no-empty-function#allow-constructors), and there is option `allow`, which accepts value `constructors` which allows to ignore `constructors`.

```TS
/* eslint no-empty-function: ["error", { "allow": ["constructors"] }]*/
class A {
    constructor() {}
}
```

@jschaf Is it acceptable for your case?

cc @bradzacher

**Comment by jschaf** [comment]
Yes, that would work.  A small downside to `constructors` is that empty public constructors are no-ops and would be ignored.

WDYT about ignoring private constructors in the code instead?  I can't think of a case where you'd want to lint an empty private constructor.  The change is backwards compatible.

**Comment by bradzacher** [comment]
could add an extra two options to the `allow` set, eg `"private-constructors" | "protected-constructors"`.
It might be a bit of a pain to implement, but I can see definite value in those within a typescript context

**Comment by zhaparoff** [comment]
And what about empty protected methods in the base class?

Let's assume the case, when we have some abstract class and a number of derived ones.
We do need some method stub in a base, because we are using it in other base class method.
But, by default it should not do anything, and only overrides can add some logic there.
Meanwhile, making it abstract will need override in every derived class, that is something I don't want to do.

```
abstract class A {
  public methodA(): void {
    this.methodB();
  }
  
  protected methodB(): void {} // <-- here the rule is triggered
}

class B extends A {
  // don't need and don't want to implement methodB here
}

class C extends A {
  private flag: boolean;

  protected methodB(): void {
    this.flag = true;
  }
}
```

Please let me know, if I need to create a separate issue for this case, but it seemed for me to be related to this discussion.

**Comment by bradzacher** [comment]
Adding an empty private/protected constructor has a tangible, semantic meaning in typescript - it means that you cannot construct an instance of the class outside of the class (or its descendants, if protected).

For your case, however, you just have an empty method...
A random empty method on a class has no semantic meaning. It's just an empty function. The fact that it's on an abstract class and that it's protected doesn't lend it any more semantic meaning than if it were a public or private empty function on a non-abstract class.

If you need an empty function implementation for your abstract protected method, then you're clearly _intentionally_ violating the lint rule, which is exactly the usecase that eslint disable comments were created for.
